src/components/Sessions/RequireSession.js

```diff
@@ -1,42 +1,113 @@
+};
+
+const RequireSession = ({expired = false, expiryDate = null, children}) => {
+  const {baseUrl} = useContext(ConfigContext);
+  const [expiryWarning, setExpiryWarning] = useReducer(
+    (draft, action) => ({...draft, ...action}),
+    {display: false, accepted: false, seconds: 0});
+
+  useEffect(() => {
+    /*
+    Show the warning if the session has expired.
+     */
+    if (expiryDate) {
+      const ms = untilWarningMilliseconds(expiryDate);
+      const timeout = setTimeout(() => {
+        setExpiryWarning({display: true});
+      }, ms);
+      return () => clearTimeout(timeout);
+    }
+  }, [expiryDate]);
 
-const RequireSession = ({ expired=false, children }) => {
-  if (!expired) {
+  useEffect(() => {
+    /*
+    Count down to the session expiry time.
+     */
+    const interval = setInterval(() => {
+      setExpiryWarning({seconds: Math.floor((expiryDate - new Date()) / 1000)});
+    }, 1000);
+    return () => clearInterval(interval);
+  }, [expiryDate]);
+
+  if (expired) {
     return (
-      <>
-        {children}
-        {/*<SessionExpiry /> disabled for now until we figure out how to style this */}
-      </>
+      <Card
+        title={<FormattedMessage description="Session expired card title" defaultMessage="Your session has expired"/>}>
+        <ErrorMessage>
+          <FormattedMessage
+            description="Session expired error message"
+            defaultMessage="Your session has expired. Click <link>here</link> to restart."
+            values={{
+              link: chunks => <Link to="/" component={Anchor}>{chunks}</Link>,
+            }}
+          />
+        </ErrorMessage>
+      </Card>
     );
+  }
+
+  const onExtendClick = async () => {
+    await apiCall(`${baseUrl}ping`);
   };
 
   return (
-    <Card title={<FormattedMessage description="Session expired card title" defaultMessage="Your session has expired" />}>
-      <ErrorMessage>
-        <FormattedMessage
-          description="Session expired error message"
-          defaultMessage="Your session has expired. Click <link>here</link> to restart."
-          values={{
-            link: chunks => <Link to="/" component={Anchor}>{chunks}</Link>,
-          }}
-        />
-      </ErrorMessage>
-    </Card>
-  );
+    <>
+      <Modal
+        title={
+          <FormattedMessage
+            description="Session expiry warning title (in modal)"
+            defaultMessage="Your session will expire soon."
+          />}
+        isOpen={expiryWarning.display && !expiryWarning.accepted}
+        closeModal={() => {
+          setExpiryWarning({accepted: true});
+        }}
+      >
+        <ErrorMessage>
+          <FormattedMessage
+            description="Session expiry warning message (in modal)"
+            defaultMessage="Your session will expire in {seconds} seconds. Extend your session if you wish to continue."
+            values={{
+              seconds: expiryWarning.seconds,
+            }}/>
+        </ErrorMessage>
```

Comment:
This should be a warning message rather than an error. No error has occurred yet, but there's one about to.